### PR TITLE
Media Store Pagination

### LINKED ISSUE #314: Re-work pagination for MediaStore
The current pagination in #274 has some issues. We decided to separate this into its own issue to better handle it.

### PR DESCRIPTION
**Pull request by oguzkocer**
Fixes #314. This PR changes how the pagination works in MediaStore. In order to make things clearer, I opted to make some changes to payloads, I also removed `MediaFilter` for now as it was not being used very much and it just made things harder to understand. While on it, I made some naming changes to better reflect how the fetch media list works.

This implementation is pretty much borrowed from PostStore and amazing job @aforcier did. Now the implementations are very close to each other, so things are a little more consistent between stores.

It may be a little more change than we planned at first, but I strongly believe this will help us in the long run. Merging this will be require some changes in wpandroid, which I am hoping to start working as soon as possible.

@aforcier Do you mind checking this one out even if @maxme would be the main reviewer? Since I followed the same patterns as you did. I'd love your feedback!

### CODE REVIEW

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 1]
```diff
@@ -30,30 +30,7 @@
...
-        public int offset;
-        public int page;
-        public SortOrder sortOrder;
-        public SortField sortField;
-        public String searchQuery;
-        public String after;
-        public String before;
-        public String mimeType;
-    }
+    public static final int NUM_MEDIA_PER_FETCH = 20;
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: What's the reasoning for changing this from `100` to `20`?

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 2]
```diff
@@ -500,11 +493,15 @@ private void performUploadMedia(MediaPayload payload) {
         }
     }
 
-    private void performFetchAllMedia(MediaListPayload payload) {
+    private void performFetchMediaList(FetchMediaListPayload payload) {
+        int offset = 0;
+        if (payload.loadMore) {
+            offset = getAllSiteMedia(payload.site).size();
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Isn't this also going to include local (non-uploaded) media that's in a pending/failed/etc state?

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 3]
```diff
@@ -565,11 +562,17 @@ private void handleMediaUploaded(@NonNull ProgressPayload payload) {
...
 
-        if (!payload.isError()) {
-            MediaSqlUtils.deleteAllSiteMedia(payload.site);
+        if (payload.isError()) {
+            onMediaChanged.error = payload.error;
+        } else {
+            // Clear existing media if this is a fresh fetch (loadMore = false in the original request)
+            // This is the simplest way of keeping our local media in sync with remote media (in case of deletions)
+            if (!payload.loadedMore) {
+                MediaSqlUtils.deleteAllSiteMedia(payload.site);
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: This was here before, but again won't this clear pending/failed media? We should only clear out `uploaded` media here.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 1 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: The default is `20`. I think the only reason we had `100` was we didn't have proper pagination and the REST API allows `100` items at the most.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 2 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Yeap, completely forgot about the states.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 3 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Yes, this should change as well!

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 1 above)
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: What about using 50 like in wpandroid `develop`?

Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Sure, I didn't give much thought on the number itself. We can change it any time. I'll change it to 50.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 4]
```diff
@@ -82,26 +59,42 @@ public boolean isError() {
...
+        }
+
+        public FetchMediaListPayload(SiteModel site, boolean loadMore) {
+            this.site = site;
+            this.loadMore = loadMore;
+        }
+    }
+
+    /**
+     * Actions: FETCH_MEDIA_LIST
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: This should be `FETCHED_MEDIA_LIST`.

**Comment by aforcier** [comment]
This looks good @oguzkocer, apart from the four comments from @maxme and myself.

I tested fetching all media in batches from the example app with Maxime's 100-image test site, all looks good 👍

After those fixes, I'll wait for a WPAndroid-side PR using this branch before merging this so we don't break `fluxc-integration` for long.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 4 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Fixed in 20510dea0bf96d8fa8d5265020421b336abb779c.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 1 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Fixed in 4d55c1e4be50f8557dbf2a29c0c48065a0012662

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 2 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Fixed in 25624f5fc7b83772886c10f7da330dafb81dcbd8

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 3 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Fixed in db1bb0b2135ea92db336ebd5d2d64b0742aa36b1

**Comment by oguzkocer** [comment]
Thanks for the review @aforcier! All fixed and will be working on the integration tomorrow!

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/MediaSqlUtils.java [hunk 5]
```diff
@@ -232,6 +240,14 @@ public static int deleteAllSiteMedia(SiteModel site) {
                 .endGroup().endWhere().execute();
     }
 
+    public static int deleteAllUploadedSiteMedia(SiteModel siteModel) {
+        return WellSql.delete(MediaModel.class)
+                .where().beginGroup()
+                .equals(MediaModelTable.LOCAL_SITE_ID, siteModel.getSiteId())
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/MediaSqlUtils.java]: This should be `siteModel.getId()` instead of `getSiteId()`.

**Comment by aforcier** [comment]
With the [id fix](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/322#discussion_r102074335) made locally, I tested:

* Fetching sites in batches, creating a local pending image in between fetches ✅
* Re-fetching remote media, local images are left untouched while existing media is wiped out and re-fetched (new local IDs) ✅

Once the ID issue is fixed + a WPAndroid-side PR, this is good to go 👍.

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/MediaSqlUtils.java (hunk 5 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/MediaSqlUtils.java]: Fixed in dedd9a51546deb32f89a86448853c7f35e645a08

**Comment by oguzkocer** [comment]
@aforcier [WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/5251) is now open, I had to make some changes in this PR, but I like how it turned out. Especially adding the `OnMediaListFetched` really removed some complication in the integration side. Let me know what you think!

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 6]
```diff
@@ -143,6 +143,21 @@ public OnMediaChanged(MediaAction cause, List<MediaModel> mediaList, MediaError
         }
     }
 
+    public class OnMediaListFetched extends OnChanged<MediaError> {
+        public SiteModel site;
+        public List<MediaModel> mediaList;
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: It doesn't look like we're using `mediaList` anywhere (including in the WPAndroid integration), except in the example apps (and that use case can be replaced by `rowsAffected` like we use in other `OnChanged` events).

It seems to me that it's better to encourage requesting the new items from the store than to pass these large lists of objects when they're not absolutely necessary.

Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: I didn't really think about that before as it was always there. But, you're right, there is no reason to pass these along. I guess that's the benefit of separating the class :)

**Comment by aforcier** [comment]
The connected tests need to be updated to use `OnMediaListFetched`.

File: example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java [hunk 7]
```diff
@@ -227,6 +227,18 @@ public void onMediaChanged(OnMediaChanged event) {
 
     @SuppressWarnings("unused")
     @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onMediaListFetched(OnMediaChanged event) {
```
Comment (aforcier) [inline example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java]: Now that this is here, we can remove the `FETCHED_MEDIA_LIST` case from `onMediaChanged`.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 6 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Fixed in e31efcdb1d84004301656bcb125089d1fd9f37df

File: example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java (hunk 7 above)
Comment (oguzkocer) [inline example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java]: Fixed in 50c29d2b7fe31b40b59ece4a8dbd9a2896a039ff

**Comment by oguzkocer** [comment]
> The connected tests need to be updated to use `OnMediaListFetched`.

Fixed in 435c827b32ab7736ef1727b018d50a429c4688f6 & c69b2221861bb5523088d2cd2df0868f3b9443a2

I've also found that we had 2 different handlers for the `onMediaChanged` event in `MediaFragment`, so I removed one of them in ba76ad87df7c597974644dfdb44d5676d9507a00

@aforcier Ready for another pass!

File: example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java [hunk 8]
```diff
@@ -231,7 +231,6 @@ public void onMediaChanged(OnMediaChanged event) {
     public void onMediaListFetched(OnMediaListFetched event) {
         if (!event.isError()) {
             prependToLog("Received successful response for media list fetch.");
-            prependToLog(event.mediaList == null ? "0" : event.mediaList.size() + " media items fetched.");
```
Comment (aforcier) [inline example/src/main/java/org/wordpress/android/fluxc/example/MediaFragment.java]: Might be useful to keep a `rowsAffected` variable so we can still have these debug logs, but I don't think that's a blocker, we can add this in the future if we feel we need to.

**Comment by aforcier** [comment]
This good to go 👍. Up to you @oguzkocer if you feel like adding a `rowsAffected` field, but it's not important.

:shipit:

**Pull request merged**